Cli key transfer (#1454)
ava-labs/avalanche-docs: docs/subnets/how-to-transfer-funds.md

> # How to Transfer Funds Between Ledgers
> 
> Occasionally, there is a need to move funds between different ledgers, or different addresses owned
> by the same ledger.
> 
> A case for the former is for a Subnet control ledger to spend all funds in previous operations,
> and, as a Subnet control ledger can't be changed, there is a need to transfer to it for another
> funded account.
> 
> A case for the later is related for example to ledger funding using web wallet transfers, which
> sometimes transfer to a ledger address associated with an index different to zero. As the CLI by
> default uses ledger index 0, sometimes a movement of the funds to that index is preferred for the
> CLI to work with the default settings.
> 
> As direct transfers between P-Chain addresses are not currently supported in wallets, CLI has added this
> command to enable them.
> 
> Internally the operation consists on a series of import/export operations involving the X-Chain also,
> and so the fee for this operation is 4 times the fee for a typical import operations, that is, 0.004
> AVAX. See [fees information](https://docs.avax.network/quickstart/transaction-fees)

Reviewer: ```suggestion
---
description: To enable direct transfers between P-Chain addresses held on a ledger,
use the Avalanche-CLI command `avalanche key transfer`.
sidebar_label: Directly Transfer Funds
---

# How to Use Avalanche-CLI to Transfer P-Chain Funds

It is sometimes necessary to transfer funds between different accounts or
addresses in our ledger. This could be due to various reasons:
1. Moving funds from a Subnet control ledger to another funded account
because the Subnet control ledger cannot be changed.
2. Transferring funds to a ledger address with a different index, which is
sometimes required for the command-line tool (CLI) to work with default
settings.

To enable direct transfers between P-Chain addresses, use the command
`avalanche key transfer`. This operation involves a series of import/export
actions with the X-Chain. The fee for this operation is four times the typical
import operation fee, which comes out to 0.004 AVAX. You can find more
information about fees [here](https://docs.avax.network/quickstart/transaction-fees).
```

Author: Will change with a slightly modified version that clarifies that the main operation is to transfer funds between different ledgers. Also fixing that the control ledger is the target of a transfer, not the source

> :::note
> 
> The `key transfer` command can also be applied to the stored keys managed by the CLI. It enables
> moving funds from one stored key to another, and from a ledger to a stored key or the other way.
> 

Reviewer: ```suggestion
:::note

The `key transfer` command can also be applied to the stored keys managed by the CLI. It enables
moving funds from one stored key to another, and from a ledger to a stored key or the other way.

:::
```

> Anyway, this tutorial focus on the most expected user case of transferring between ledger accounts.

Reviewer: ```suggestion
This tutorial focuses on transferring funds between ledger accounts.
```

best practice to simplify and remove casual language

## Prerequisites

- [`Avalanche-CLI`](https://github.com/ava-labs/avalanche-cli) installed
- Multiple Ledger devices [configured for Avalanche](./create-a-mainnet-subnet#setting-up-your-ledger)

## Example: Sending All Funds From One Ledger to Another

- Source address: ledger A, index 2 (the web wallet shows 4.5 AVAX for this ledger)
- Target address: ledger B, index 0 (the web wallet shows 0 AVAX for this ledger)

> ### Decide on Which Index of Ledger A to Use

Reviewer: ```suggestion
### Determine the Address Index
```

> It is needed to find out which index is being used by the web wallet. Say that the web wallet
> shows 4.5 AVAX available on p-chain address P-avax10an3cucdfqru984pnvv6y0rspvvclz63e523m0
> 
> With the ledger A connected and the avalache app running, execute:
> 
> ```bash
> avalanche key list --mainnet --ledger 0,1,2,3,4,5
> ```
> 
> To see p-chain addresses and balances for the first 6 indices in the ledger derived owner addresses.
> 
> ```text
> +--------+---------+-------------------------+-----------------------------------------------+---------+---------+
> |  KIND  |  NAME   |          CHAIN          |                    ADDRESS                    | BALANCE | NETWORK |
> +--------+---------+-------------------------+-----------------------------------------------+---------+---------+
> | ledger | index 0 | P-Chain (Bech32 format) | P-avax1g8yucm7j0cnwwru4rp5lkzw6dpdxjmc2rfkqs9 |       0 | Mainnet |
> +        +---------+                         +-----------------------------------------------+---------+---------+
> |        | index 1 |                         | P-avax1drppshkst2ccygyq37m2z9e3ex2jhkd2txcm5r |       0 | Mainnet |
> +        +---------+                         +-----------------------------------------------+---------+---------+
> |        | index 2 |                         | P-avax10an3cucdfqru984pnvv6y0rspvvclz63e523m0 |     4.5 | Mainnet |
> +        +---------+                         +-----------------------------------------------+---------+---------+
> |        | index 3 |                         | P-avax1yfpm7v5y5rej2nu7t2r0ffgrlpfq36je0rc5k6 |       0 | Mainnet |
> +        +---------+                         +-----------------------------------------------+---------+---------+
> |        | index 4 |                         | P-avax17nqvwcqsa8ddgeww8gzmfe932pz2syaj2vyd89 |       0 | Mainnet |
> +        +---------+                         +-----------------------------------------------+---------+---------+
> |        | index 5 |                         | P-avax1jzvnd05vsfksrtatm2e3rzu6eux9a287493yf8 |       0 | Mainnet |
> +--------+---------+-------------------------+-----------------------------------------------+---------+---------+
> ```
> 
> As can be noticed, the address `P-avax10an3cucdfqru984pnvv6y0rspvvclz63e523m0` indeed has 4.5 AVAX and
> is associated to index 2 of the ledger A.
> 
> :::note
> 
> A ledger manages an infinite amount of addresses derived from a main private key, and due to this,
> in certain operations there is a need to specify which address or which index of the ledger is the user
> operating against.
> 

Reviewer: ```suggestion

A ledger can manage an infinite amount of addresses derived from a main private key. Due to this,
many operations require an address index the user wishes to use in an operation

After confirming with a web wallet that 4.5 AVAX is available on p-chain address 
`P-avax10an3cucdfqru984pnvv6y0rspvvclz63e523m0`, connect ledger A.

With the avalanche app running, execute:

```bash
avalanche key list --mainnet --ledger 0,1,2,3,4,5```

To see p-chain addresses and balances for the first 6 indices in the ledger derived owner addresses.

```text
+--------+---------+-------------------------+-----------------------------------------------+---------+---------+
|  KIND  |  NAME   |          CHAIN          |                    ADDRESS                    | BALANCE | NETWORK |
+--------+---------+-------------------------+-----------------------------------------------+---------+---------+
| ledger | index 0 | P-Chain (Bech32 format) | P-avax1g8yucm7j0cnwwru4rp5lkzw6dpdxjmc2rfkqs9 |       0 | Mainnet |
+        +---------+                         +-----------------------------------------------+---------+---------+
|        | index 1 |                         | P-avax1drppshkst2ccygyq37m2z9e3ex2jhkd2txcm5r |       0 | Mainnet |
+        +---------+                         +-----------------------------------------------+---------+---------+
|        | index 2 |                         | P-avax10an3cucdfqru984pnvv6y0rspvvclz63e523m0 |     4.5 | Mainnet |
+        +---------+                         +-----------------------------------------------+---------+---------+
|        | index 3 |                         | P-avax1yfpm7v5y5rej2nu7t2r0ffgrlpfq36je0rc5k6 |       0 | Mainnet |
+        +---------+                         +-----------------------------------------------+---------+---------+
|        | index 4 |                         | P-avax17nqvwcqsa8ddgeww8gzmfe932pz2syaj2vyd89 |       0 | Mainnet |
+        +---------+                         +-----------------------------------------------+---------+---------+
|        | index 5 |                         | P-avax1jzvnd05vsfksrtatm2e3rzu6eux9a287493yf8 |       0 | Mainnet |
+--------+---------+-------------------------+-----------------------------------------------+---------+---------+```

The address `P-avax10an3cucdfqru984pnvv6y0rspvvclz63e523m0` has 4.5 AVAX and
is associated to index 2 of ledger A.

```

> ### Decide which Index of Ledger B to Use and Find Out the Associated Address

Reviewer: These subtitles appear in the sidebar on the right-hand side of the page and are easily truncated when they are too long. 
```suggestion
### Determine Receiver Address Index
```

obtain it:

With the ledger B connected and the avalache app running, execute:

```bash
avalanche key list --mainnet --ledger 0
```

```text
+--------+---------+-------------------------+-----------------------------------------------+---------+---------+
|  KIND  |  NAME   |          CHAIN          |                    ADDRESS                    | BALANCE | NETWORK |
+--------+---------+-------------------------+-----------------------------------------------+---------+---------+
| ledger | index 0 | P-Chain (Bech32 format) | P-avax1r4aceznjkz8ch4pmpqrmkq4f3sl952mdrdt6xm |       0 | Mainnet |
+--------+---------+-------------------------+-----------------------------------------------+---------+---------+
```

Target address to be used is `P-avax1r4aceznjkz8ch4pmpqrmkq4f3sl952mdrdt6xm`, containing 0 funds.

### Execute the Sending Part of the Transfer Operation

A P-Chain to P-chain transfer is a two-part operation. There is no need for the two parts to be executed
on the same machine, only for them to have some common params. For each part, the appropriate ledger
(either source or target) must be connected to the machine executing it.

> First part starts the process, and moves the money out of the

Reviewer 2: ...into an X-Chain account owned by the receiver.

source account into a X-Chain account owner by the receptor. It needs to be signed by the sending ledger.

> The amount to be specified as input to the command is the amount to be received on the target address.

Reviewer 2: Enter the amount of AVAX to send to the recipient. This amount does not include fees.

> As in this example the users wants to send all the balance of 4.5, so he needs to consider a fee payment

Reviewer 2: This contradicts the previous sentence. Should you include fees in the entry or not?

Author: The entry does not include fees. If the user wants to send all available money, he should consider the total balance, substract the fee, and ask the command to send the result of that substraction.
I will remove this explanation as it confuses more than it clarifies.

of 0.004, and tell the command to transfer 4.496.

> Take into account that the sending ledger is the one that is going to pay all the fees.

Reviewer 2: `Take into account` -> `Note`. `Is the one that is going to pay` -> `pays`

```

Next, the step of the transfer must be specified. Send in this case:

```text
? Step of the transfer:
  ▸ Send
    Receive
```

Next, the key source for the sender address. That is, the key that is going to sign the sending
transactions. Select `Use ledger`:

```text
? Which key source should be used to  for the sender address?:
    Use stored key
  ▸ Use ledger
```

Next, the ledger index is asked for. Input `2`:

```text
✗ Ledger index to use: 2
```

Next, the amount to be sent is asked for:

```text
✗ Amount to send (AVAX units): 4.496
```

The, the target address is required:

```text
✗ Receiver address: P-avax1r4aceznjkz8ch4pmpqrmkq4f3sl952mdrdt6xm
```

After that, a confirmation message is printed. Read carefully and choose `Yes`:

```text
this operation is going to:
- send 4.496000000 AVAX from P-avax10an3cucdfqru984pnvv6y0rspvvclz63e523m0 to target address P-avax1r4aceznjkz8ch4pmpqrmkq4f3sl952mdrdt6xm
- take a fee of 0.004000000 AVAX from source address P-avax10an3cucdfqru984pnvv6y0rspvvclz63e523m0

Use the arrow keys to navigate: ↓ ↑ → ←
? Confirm transfer:
    No
  ▸ Yes
```

After this, the first part is completed:

```text
Issuing ExportTx P -> X
```

> ### Execute the Receiving Part of the Transfer Operation

Reviewer 2: Receive the Transfer

> This part of the operation, to be signed by ledger B, is going to take the funds from a X account owned

Reviewer 2: In this step, Ledger B signs the transaction to receive the funds. It imports the funds on the X-Chain before exporting them back to the desired P-Chain address.

by B, and move them to the expected P-Chain address.

Connect ledger B and execute avalanche app.

Then start the command:

```bash
avalanche key transfer
```

Specify the `Mainnet` network:

```text
Use the arrow keys to navigate: ↓ ↑ → ←
? Network to use:
  ▸ Mainnet
    Fuji
    Local Network
```

Next, the step of the transfer must be specified. Receive in this case:

```text
? Step of the transfer:
    Send
  ▸ Receive
```

Then, select Ledger as the key source that is going to sign the receiver operations.

```text
? Which key source should be used to  for the receiver address?:
    Use stored key
  ▸ Use ledger
```

Next, the ledger index is asked for. Input `0`:

```text
✗ Ledger index to use: 0
```

Next, the amount to receive is asked for:

```text

> ✗ Amount to send (AVAX units): 4.496

Reviewer 2: Note, we should add a CLI ticket here to add a `MAX` option
